### ISSUE TITLE
Does API.mli serve any purpose and what do we do about it?

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#6008]
By reverting the API in v8.7 we have been able to delay this discussion and at the last WG there was already too much to be discussed, but this question should be dealt with anyways.

From reading Gitter, one gets the impression that no one really knows what the API is for.

Is it an API in the proper sense? Is it designed to be the restricted set of functions to which plugin writers have access? To me the answer to this question is clearly yes, but @maximedenes reminds that the API contains a lot of stuff that is not recommended to use, while it is missing a lot of the proper alternatives. This clearly sends a bad signal and given that plugins go through the API *by default* (which I protested against, fact that I will continue to remind), we really *have to* do something about it.

@maximedenes seems to think that it serves another purpose that is to tag which functions are used in plugins. This may be fine for Coq developers developing official plugins, but opening a PR every time you need a function that is not exported when you are an external plugin writer is not.

What should we do about it? I can see several possibilities:

1. Remove the API entirely. Its author is not in the team anymore anyways and it looks like no one wants to do the huge amount of work that it is badly needing. Then improve step-by-step our internal APIs by deprecating and removing stuff, uniformizing and simplifying names and, most importantly, documenting stuff. We can also discuss trying to restrict the API that is exported component by component (but then the other components can't bypass it).
2. Disable the API by default until we make it better (but honestly, I doubt this will ever happen with the current approach).
3. Remove the huge mess that is the current `API.mli` and immediately create a much smaller one which is properly crafted. All plugins will need to bypass it at start because they will all be missing lots of stuff, but we can then try to make them not bypass it one by one (by adding required functions to `API.mli` when it makes sense to export these functions, and by removing the use of functions that shouldn't be exported). Because this will proceed by much smaller PRs, this will allow for a proper review process, while the current API was never properly reviewed (and that's why it created so many problems).

**Comment by maximedenes** [comment]
IMHO We should either have a plan for which this "API" serves some purpose, or remove it. So I'd be strongly against 2.

**Comment by ejgallego** [comment]
> From reading Gitter, one gets the impression that no one really knows what the API is for.

The word "API" has been used in the context of Coq for many different things, but for the piece of code the goals were very concrete: Coq's "API" was created to isolate the ML-interface used by plugins.

The first iteration of that interface included all the functions actually used by the plugins, using a top-level module to implement it.

In my opinion, the whole effort seemed a bit strange from the start, first, it is not clear why "plugins" ought to have a different API than the rest of the system. I would think that components such as the pretyper, or the document manager (STM) would expose the same API to all their users indistinctly.

Second, we often see the confusion and complains about the current technical solution used to implement the "API" interface, which indeed could be easily improved, and actual question of code organization and design.

Third, I think we have been selling the view of "let's make an API" as a kind of deliverable, but however API-design reality is far more subtle and dynamic than that. In real world systems APIs are rarely "set in stone" but they do evolve and coexists with competing APIs for a long time. It is often more practical in large projects to maintain a large coherent repository than many smaller ones, specially if rapid development is desired, or in Coq's case, coupling is fundamentally high. API stability seems like an utopia given what the current status is.

A possible direction could be what we have discussed some times: see Coq as a library composed of different modules (see META for a start), and work towards making each module an independent unit, without losing the global coherence == all the modules live in the same repository and coupling can still be high, and API stability can still be low.

**Comment by Zimmi48** [comment]
> In my opinion, the whole effort seemed a bit strange from the start, first, it is not clear why "plugins" ought to have a different API than the rest of the system. I would think that components such as the pretyper, or the document manager (STM) would expose the same API to all their users indistinctly.

:+1:

**Comment by ejgallego** [comment]
A big todo in the API front is to clean up the interdependency mess wrt `engine` and `library` and `intf`

**Comment by SkySkimmer** [comment]
Is there a way to observe these interdependencies?
Does #1033 make it worse? (engine/universes calling into library/lib)

**Comment by Zimmi48** [comment]
There are technical questions as well, but they probably deserve their own issues. This one is really to discuss the purpose of the "API" infrastructure itself.

**Comment by ejgallego** [comment]
@SkySkimmer you can have a look at the diff of https://github.com/ejgallego/coq/tree/jbuilder I guess I will port some of that stuff.

**Comment by ejgallego** [comment]
See also #6015

**Comment by ejgallego** [comment]
> There are technical questions as well, but they probably deserve their own issues. This one is really to discuss the purpose of the "API" infrastructure itself.

I am not sure you can really separate the two.

**Comment by mattam82** [comment]
I think the current API shows that we have an organically developped API to start with and that a considerable amount of work is needed to curate it and put it in usable form. Whether we do this bottom-up from smaller components or top-down by refining this big block doesn't really matter to me, but of course this top down approach is putting in front of everyone's eyes the global state of affairs. I guess we should answer the question if we want to invest, about right now, time on making a usable, comprehensive public interface. Personally I'm not too excited by the subject as a whole, but I'm willing to spend time on making the interfaces I'm responsible for cleaner. I will test compiling Equations without bypass-API to see if the API at least already serves the purpose of delimiting "public" functions.

**Comment by Zimmi48** [comment]
> I guess we should answer the question if we want to invest, about right now, time on making a usable, comprehensive public interface.

@coq/core Who answers "Yes, I want to invest time, right now, on making a usable, comprehensive public interface."?

If no one or almost no one does, we can't keep this big mess. The top-down refinement approach is not practicable if the API stays in such a state for too long.

If people are not willing to invest lots of time right now but are OK to spend some time on a longer term project of "making a usable, comprehensive public interface", then the effort must be bottom-up and we need to organize it.
Then the first question to be answered is:

> why "plugins" ought to have a different API than the rest of the system. I would think that components such as the pretyper, or the document manager (STM) would expose the same API to all their users indistinctly.

@coq/core Once again a question to answer: who thinks what?

If most people think that plugins should have a different API, then this is solution 3 of the issue description: create a new, almost empty `API.mli` and chose what to export and adapt plugins one by one to use only what is exported.

If most people think that each component (which roughly correspond to directories in the source code) should export a restricted set of functions for use both in the rest of the Coq source code, and in plugins, then we should create a sort of `API.mli` per component.
This presents the advantage that if no-one is ready to invest time on building the API for a specific component right now, we don't need to create the API for this component, but this won't prevent progress being made on other components.

**Comment by ejgallego** [comment]
@Zimmi48 I am very interesting on improving the API of a few parts of the system in particular as I use them in other external projects and I've found they need improvement.

Regarding the `API` .mli file by itself, I am not sure that is the best technical approach, but I think we should hold a call discussion several options that OCaml's module system can give to us here.

**Comment by ppedrot** [comment]
As for myself:
- I was believing at first that the API as developed by @matejkosik would be a very good, urgent thing to do. Unluckily I got brutally disillusioned for various reasons, including:
  1. The initial inclusion criterion is completely wrong, Coq-distributed plugins are completely out of phase with the current practices and we don't have the manpower to port them. This resulted in deprecated-but-we-don't-have-the-right-to-say-it code being included and their proper replacement not. In particular no way to write correct code without bypassing the API.
  2. Making the API module-bound is a complete anti-feature and makes the life of external plugin developers a terrible hell. At least I speak for myself. As @ejgallego mentions, I don't know if this is solvable by some dedicated tools or an essential limitation of the module system.
  3. Code duplication. Really.
- I'm not willing to spend energy in the API as-is, as I don't think we can solve it this way. It was a good failed experiment, as it showed the limitations of the approach.
- I do want to provide some form of API though, and I'd like to participate on such an effort. What I consider to be a reasonable alternative way would be to tag functions and structures from the mli files with a dedicated tag (e.g. using PPX) so that it shouts a warning at users if they use it. So, no OCaml modules, external-tool based very lightweight system that does not harass people knowing what they are doing.

**Comment by herbelin** [comment]
I basically agree with the various comments above. I have a question in the short term. While adding a function, I added a dependency of `Geninterp` (`engine`) into `Tactypes` (`intf`) and I realized that I was falling on the following: "A big todo in the API front is to clean up the interdependency mess wrt engine and library and intf". Indeed, some of `intf` depends on `engine` which depends on another part of `intf`. @ejgallego: would splitting `intf` in two parts be enough, or is there more to do?

In particular, my recent DX (developer experience) is that merging changes in `API` is not the most exciting experience to have, namely because of conflicts in blocks of declarations moved around. But if we arrive to an `API` which strictly follows the effective order of dependencies and whose structure remains globally stable, wouldn't it be easier to merge?

**Comment by ejgallego** [comment]
@herbelin I did some more progress on that front [but didn't update that comment]. I believe that now `intf` is now free of such dependency after bf4112094feb1a705d8bdaea3fb0febc4ef3ff59 , so you may try moving things around.

**Comment by ejgallego** [comment]
Ok, I've checked my internal tree modularizing Coq with jbuilder, and indeed I think that all that remains is to move/refactor `Tactypes`. I placed `Tactypes` in `interp`, but it could go as well in `tactics`, with a bit of effort.

**Comment by ejgallego** [comment]
In fact types such as `glob_constr_and_exprp` should likely be moved to `Stdarg` and the rest to tactics?

**Comment by ejgallego** [comment]
I'd be great if we get rid of `Tactypes` I think.

**Comment by ejgallego** [comment]
@herbelin see https://github.com/coq/coq/pull/6052

**Comment by herbelin** [comment]
Thanks for #6052!

**Comment by ejgallego** [comment]
So I've ported three reasonably representative plugins `unicoq/mtac2/paramcoq` to 8.8 and here are my thoughts:
- Having `API.mli` is painful, but it serves a role for now. Plugins are really using crazy stuff, and it is good to have to track it explicitly.
- The amount of redundancy the base Coq interfaces have is really large. Not only for types, but also for functions. Example, `start_module`. What is worse, some plugins call `Lib.start_module` and close it uising `Declare_mod.close_module`, etc..
- Plugins tend to copy things that are done in main codebase. In many cases, we do crazy/outdated/wrong things in the main codebase, so that's propagated to plugins.
- Documentation on how to use the functions / perform basic tasks is non existent or just misleading [with a few notable exceptions] `dev/doc/changes` was very far from complete.

So indeed there is lot of hard work to be done here. IMHO it should be project-wide push to improve this situation, and it will take a long time. Also, it is clear we want to change the technical mechanism used in `API.mli` to duplication-free approach. Some ideas:

- We should make a list of "essential" operations, such as "how to declare a constant", "how to typecheck a term", etc... and make them available in the developer reference.
- We need to reduce the amount of redudancy, and we need to refactor some parts of the main codebase following the above guidelines.
- We should require better documentation before merging PRs that modify interfaces.

With regards to the technical solution, I am leaning towards an approach in which every component (let it be the kernel or even finer) provides two modules `Public` `Internal`, and we use module aliases to export it. For example, modules within `kernel` can see the `Internal` interfaces, modules outside it should open it and are thus "tagged". WDYT?

What I propose is that we start implementing this approach incrementally. Let's move the modules in `kernel` from `API.mli` to this approach, and let's see how it goes. Obviously this approach needs some careful design. Coincidently, `kernel` is where the biggest mess due to the `API.mli` has occurred due to inconsistent deprecation policy, so that should be a good move too.

**Comment by herbelin** [comment]
I don't remember if this statistics was done. First number is what is exported in `API`, second is what is exported in the files of the corresponding directory (for a branch of two days ago).
```
   config:                   1/  51 functions
      lib:   0/ 157 types    0/ 924 functions
   kernel: 177/ 271 types  387/1487 functions
  library:  31/  62 types  160/ 486 functions
     intf: 179/ 210 types
   engine:  57/  96 types  334/ 823 functions
pretyping:  41/  88 types  132/ 628 functions
   interp:  24/  42 types  105/ 303 functions
   proofs:  21/  28 types  110/ 292 functions
  parsing:  16/  23 types   96/ 116 functions
 printing:   2/   8 types   56/ 181 functions
  tactics:  33/  61 types  278/ 571 functions
   vernac:  13/  22 types   49/ 214 functions
      stm:   1/  35 types    7/ 148 functions
 toplevel:   0/   1 type     0/  26 functions
total API: 595/1104 types 1715/6250 functions
```

**Comment by ejgallego** [comment]
Interesting, thanks @herbelin , I think the stats support the usefulness of the current approach. [note that I had to add a few dozen functions to the API to get the above plugins to compile]

I forgot to mention a last point above, validation. I will edit the post.

**Comment by ejgallego** [comment]
Ok, so a small technical brainstorming on how to move forward with `kernel`.

What we are going to do is to create a new module `Kernel_API` [1], that will contain all the modules from `kernel` with it's public interfaces plus an `Internal` module, exposing these parts for those looking for trouble.

So each module in `kernel` keeps the same implementation, but interfaces are structured as:
```ocaml
(* names.mli *)
module type Public = sig
   ....
end

module type Internal = sig
   ....
end
```
How to better create the implementation `names.ml` is indeed an open question.

Once we have that for each module, `Kernel_API.mli` does:
```
include module type of struct include Names.Public
include module type of struct include Constr.Public
....

module Internal = sig
   include module type of struct include Names.Internal
   ...
end
```

with the implementation just defining the proper module aliases. There are still a few technical details to solve but I hope you get the idea. WDYT?

[1] Should we use `jbuilder` this would be named `Kernel`.

**Comment by maximedenes** [comment]
I think the API plan would deserve a summary, like in a CEP or something, so that we can get the full picture and say how we feel about it before starting concrete action.

**Comment by ejgallego** [comment]
> I think the API plan would deserve a summary, like in a CEP or **something**,

Such as its own issue?

**Comment by maximedenes** [comment]
The key word was "summary". I meant a summary of the agreed plan of action, that wouldn't require to go trough the entire discussion, if it seems to be converging.

**Comment by ejgallego** [comment]
I do however think that what the API (and in general Coq) needs is **more concrete action** not less. The ratio of code to discussion is high enough as it is. Discussion sure is useful, but code is where the real deal is.

**Comment by ejgallego** [comment]
> I meant a summary of the agreed plan of action,

There is no agreed plan of action AFAICS. I would be happy if for once this issue would focus more on actual technical discussion and less on management.

**Comment by maximedenes** [comment]
> Having API.mli is painful, but it serves a role for now. Plugins are really using crazy stuff, and it is good to have to track it explicitly.

I don't understand this. Like PMP said, what we track is the stuff that was used at some point by some plugin, and may still be or not. What do you plan to use that information for?

**Comment by maximedenes** [comment]
Maybe I misunderstood what you are about to do. What do you plan to put in each component's public interface? A set of functions that make sense or the random stuff that is in `API.mli` today?

**Comment by ejgallego** [comment]
> What do you plan to use that information for?

It is a starting to point. For example, when eviscerating `kernel` from `API.mli`, I think what it is exposed there is a good base to define `Kernel.Public` [or just `Kernel`].

Of course, a second step is curating `Kernel.Public` so that it makes sense.

**Comment by ejgallego** [comment]
> I misunderstood what you are about to do.

I am not sure I/we are going to do this. I am just trying to propose a path to move forward. `rm -rf API` is _not a plan foward_ for me, but a step backwards.

Other proposals are welcome. As I have said, I don't consider `rm -rf API` an acceptable plan.

**Comment by SkySkimmer** [comment]
> 
> So each module in `kernel` keeps the same implementation, but interfaces are structured as:
> ```ocaml
> (* names.mli *)
> module type Public = sig
>    ....
> end
> 
> module type Internal = sig
>    ....
> end
> ```

Doesn't that mean every use of some value will be like `Foo.Public.bar`? That sounds quite painful.

>How to better create the implementation names.ml is indeed an open question.

The best I can see is something like
``` ocaml
(* foo.ml *)
module Internal = struct
  let bar = ...
end
module Public = Internal
```

**Comment by maximedenes** [comment]
I agree with the idea of writing a public interface for the kernel, if it doesn't make it heavier for external components to use it. If you think it is easier to start from `API.mli`, then fine (I'm not convinced it is).

**Comment by ejgallego** [comment]
In a sense I am proposing a concrete technical realization of @Zimmi48 's original point no. 3.

> Doesn't that mean every use of some value will be like Foo.Public.bar? That sounds quite painful.

Indeed, we need to tweak that. For the people opening `Kernel_API` it wouldn't be the case, but there is still the question of intra-kernel use... Maybe we would need to add `module Name = Name.Public`. I am not sure, just really brainstorming for now. This is not an easy problem. On the other hand we shouldn't make `Internal` be a submodule of `Public` I think, but maybe we can do better like that.

Note that `jbuilder` will indeed `open Kernel_API` automatically when you say that your component uses the `kernel` package.

> ```ocaml
> (* foo.ml *)
> module Internal = struct
>   let bar = ...
> end
> module Public = Internal
> ```

That looks good to me.

**Comment by ejgallego** [comment]
> If you think it is easier to start from API.mli, then fine (I'm not convinced it is).

Well, it provides a reference point --- better or worse --- as to what to put in `Kernel.Public`, a lot of work was already done, and I am skeptical we will ever redo it.

I am --- slightly only --- more hopeful that if for example, `Names.mli` contains both a public and internal part, people would submit "curating" PRs.

**Comment by maximedenes** [comment]
> I am --- slightly only --- more hopeful that if for example, Names.mli contains both a public and internal part, people would submit "curating" PRs.

But can we do so, as long as we don't port plugins to non deprecated stuff (at least the ones shipped with Coq)?

**Comment by maximedenes** [comment]
(I'm just asking how you see the sequence of concrete actions here)

**Comment by ejgallego** [comment]
I think a possibility is:

- design a system that allows to expose a public interface and doesn't need the current duplication. Example, the above idea using module, module types, etc... This needs a lot of design work.
- gradually migrate components out of API.mli to the new, duplicate-free system. At the same time consolidate the current mess of deprecation, and adjust interfaces if possible.
- individually curate the public / internal / deprecated parts of each component.

Porting is IMO a bit more orthogonal, as deprecation implies that nothing in core will compile with `-warn-error`. So people introducing the deprecation should do it.

**Comment by maximedenes** [comment]
> Porting is IMO a bit more orthogonal, as deprecation implies that nothing in core will compile with -warn-error. So people introducing the deprecation should do it.

But my point is that our plugins today use a lot of already deprecated stuff. So either you port the plugins, or you add the already deprecated stuff in the API, or you make those plugins bypass the API. Which option do you suggest?

**Comment by ejgallego** [comment]
I don't think we can control that. Obviously the best is to clean up the plugins, but in the absence of manpower, what do you suggest? If these plugins have to `open Internal`, such is life.

The rule for deprecated stuff [that is marked with `ocaml.deprecated`] is clear, the person introducing the deprecation has to take care of the porting.

What I am focused now is in a concrete solution that avoids the current pain with duplication of interfaces. I think your concern is fairly orthogonal to what I'm trying to do.

**Comment by maximedenes** [comment]
> If these plugins have to open Internal, such is life.

Yeah, I think that's the right approach. This way, plugins that haven't been ported won't impact the API draft, and the problems will indeed by orthogonal.

But notice that it is not at all what we did so far with `API.mli`, where we added a lot of things coming from non-ported plugins. So we may want to remove that from the `Public` interface as an immediate second step after the resolution of duplication.

**Comment by ejgallego** [comment]
So I wrote some code and for now I see two technical problems with my proposal:

- The current build system is not well adapted to express the dependencies of components, for example, `pretyping` should get `-open Kernel_API -open Engine_API` etc... but that requires some Makefile surgery :S Both Ocamlbuild and JBuilder can do this in an easier way.
- It is convenient to have the `Public` and `Internal` modules for each sub-component separate, so leaks occur, however this is a bit painful due to having to add a few `module Foo = Name.Public` in the files.

> Yeah, I think that's the right approach. This way, plugins that haven't been ported won't impact the API draft, and the problems will indeed by orthogonal.
> 
> But notice that it is not at all what we did so far with API.mli, where we added a lot of things coming from non-ported plugins. So we may want to remove that from the Public interface as an immediate second step after the resolution of duplication.

Indeed, I fully agree. The first iteration of the API didn't mark the bad stuff, this is why I think we need the `Public` vs `Internal` distinction. Now, it is just a matter of how smart we can get with module hackery to have a nice implementation of the idea.

**Comment by ejgallego** [comment]
Another choice is indeed to declare a module type for each module, and then use include to avoid duplication:

```ocaml
(* names_intf.ml *)
module type Public = sig
   val foo ...
end

module type Internal = sig
   val foo...
end

(* names.mli *)
include Names_intf.Public
module Internal : Names_intf.Internal

(* names.ml *)
module Internal = struct
   let foo ...
end
include Internal
```

But that looks a bit clunky too...

**Comment by SkySkimmer** [comment]
I don't see what that gives you compared to the other system.

The main issue is that in order to avoid code duplication we want to be able to `include` or alias all of Name, so we have to put its entire implementation inside an internal module which is ugly.

**Comment by ejgallego** [comment]
> I don't see what that gives you compared to the other system.

`open Name` works for users in `kernel/`, however the other system may work too.

> The main issue is that in order to avoid code duplication we want to be able to include or alias all of Name, so we have to put its entire implementation inside an internal module which is ugly.

Well, we could certainly place the boundaries at the "directory" level, and have `Kernel_API` do the alias work, however this would mean that uses of lets say, `Name` inside `kernel/` won't get access to the public/private machine, right?

**Comment by SkySkimmer** [comment]
>open Name works for users in kernel/, however the other system may work too.

Why have the inft file instead of having it inlined in the mli?

>Well, we could certainly place the boundaries at the "directory" level, and have Kernel_API do the alias work, however this would mean that uses of lets say, Name inside kernel/ won't get access to the public/private machine, right?

That doesn't sound like a problem to me TBH.

**Comment by ejgallego** [comment]
> Why have the inft file instead of having it inlined in the mli?

I think you are right so my current iteration is:

```ocaml
(* names.mli *)
module Public : sig
   val foo ...
end

module Internal : sig
   val foo...
end
include module type of Public

(* names.ml *)
module Internal = struct
   let foo ...
end
module Public = Internal
include Public
```

**Comment by ejgallego** [comment]
> That doesn't sound like a problem to me TBH.

I disagree here, for the reasons @maximedenes outline above, I think we want to track which parts of `kernel/` itself use other `Internal` components. However the above solution does that OK without too much overhead I think. [I am not sure how bad `include` is]

**Comment by ejgallego** [comment]
However a problem with the current solution is that it is easy for modules to skip `Names.Public`, ummm, dunno where to enforce that.

**Comment by ejgallego** [comment]
@maximedenes here is a first-first task: remove `Names` from `API.mli`

How should we proceed? The interfaces of `Names` and `API.Names` do differ significantly. Should we go ahead and incorporate all the deprecation that happened in `API.Names` into `Names`? Is there some internal part for such a module?

**Comment by ejgallego** [comment]
In particular, is there consensus as to prefer `Constant.t` to `constant`, `Constr.t` to `constr`, etc... ?

@herbelin ?

**Comment by SkySkimmer** [comment]
> 
> Once we have that for each module, `Kernel_API.mli` does:
> ```ocaml
> include module type of struct include Names.Public
> include module type of struct include Constr.Public
> ....
> 
> module Internal = sig
>    include module type of struct include Names.Internal
>    ...
> end
> ```

How does this work with deprecation? ie if I deprecate something in names.mli does it produce the warning for people using it through kernel_api? Does it produce a warning when compiling kernel_api?

**Comment by ejgallego** [comment]
> How does this work with deprecation? ie if I deprecate something in names.mli does it produce the warning for people using it through kernel_api? Does it produce a warning when compiling kernel_api?

I think it should preserver the attributes, at least in recent OCaml versions but we should test it. Such stuff has changed in 4.06 for example.

Anyways, I think such module is basically redundant and we may move forward without it first [but indeed at some point we should ]

**Comment by herbelin** [comment]
> In particular, is there consensus as to prefer Constant.t to constant, Constr.t to constr, etc... ?

It seems to me that this is the current "norm". This is how @ppedrot made a lot of restructuration work in the past years, this is how we accepted #561. I also talked e.g. yesterday with @letouzey and this is also his credo. So, consistency in our actions would require that we go in this direction.

The only possible worry I see is the compatibility, or maybe more precisely the incrementality of incompatibilities, but this is a side question about when is the alias deprecated and dropped.

As for `Constr.t` specifically, I like the idea of `types` as a documentation of whether a function expects an arbitrary term or a term which is a type. I don't how much this distinction is appreciated in general, but, if kept, there could probably be a `Type.t`.

My 2p.

**Comment by ejgallego** [comment]
Thanks for the feedback @herbelin , indeed we should be sure we are all in the same boat here IMHO as merging the `Names` and `API.Names` interfaces will be a lot of work, the diff size will be very large if we want to deprecate all the aliases.

**Comment by ejgallego** [comment]
Ok I've submitted #6064 to move the discussion about uniform alias reduction there.

**Comment by ejgallego** [comment]
Ok, so first prototype implemented in #6066 , let me know what you thinks folks; hopefully we can move on with this soon.

**Comment by ejgallego** [comment]
Another proposal, if we deem #6066 or some variation of it OK, then the main task is to move things from `API` back to the public part of their modules.

So indeed as @maximedenes rightly points out, just putting what's in API in the public part is a very bad idea. So I suggest we do one PR for each module, and use the PR to review _what_ should go in the public part.

The good part of this is that I think it is a full incremental system, if we do 20 modules for 8.8, then so be it.

**Comment by ejgallego** [comment]
Ok, #6067 can serve as a discussion of `Internal`

**Comment by ejgallego** [comment]
Ok, so I did a lot of progress on investigating possibilities using the module system and in the general deprecation of aliases; `Term/Constr` is in particular a bummer, IMHO `Constr` should have never been introduced in the first place but the least disruptive now is just to replace `open Term` by `open Constr`.

There are several possibilities in order to have a `Public` / `Internal` interface for a module `A`:
- Declare the "Public" interface as the top level interface, have an `Internal` module. This is simple but has the disadvantage that users of `Public` may potentially refer to `Internal` implicitly.
- Declare a `Public` and and `Internal` interfaces in the top-level of `A`, then we can include `Public` in the top-level of `A`, with or without equalities. This allows interfaces re-exporters to narrow down to the `Public` part.

Unfortunately OCaml's 4.02.3 `include` is quite broken, so some of these solutions would require raising the minimal version to `4.03.0`, this would brings some other benefits, but dunno if enough.

**Comment by ejgallego** [comment]
Ok so I think some parts of this are mature enough as to proceed. Now @maximedenes it is you who has to tell us about your roadmap so we can proceed to integrate things, for as you can figure out doing this uncoordinated would be hell due to the size of some commits. I think there are 3 fairly orthogonal [but interdependent] tasks:

- work on module structure organization (example #6052). There are still some other changes needed, for instance `Decl_kinds` is part of the kernel and so should be moved. `Evar_kinds` also gets a dependency from `Globnames` and that is unfortunate... wondering how to solve that. These kind of commits are small and pose no significant problem, but clear instructions on the timeframe should be provided as not to waste work.
- work on transferring "deprecation" from `API.mli` to core. This is mostly done, but again needs a specific integration timeframe.
- work on access control / removing API redudancy. This should happen after the previous points are solved, and poses no major difficulty I think, other than deciding what/if the current technical solution is OK.

**Comment by SkySkimmer** [comment]
The Public/Internal thing loses the information of what functions are used by plugins in favour of giving them all Public stuff doesn't it?
Would we make any promises of stability in the Public part or something? I still don't really understand what goes in Internal beyond renaming Unsafe to Internal.

>Evar_kinds also gets a dependency from Globnames and that is unfortunate... wondering how to solve that.

Merge #6092, remove `open Globnames`, done ;)

**Comment by ejgallego** [comment]
> Merge #6092, remove open Globnames, done ;)

Umm, how does #6092 remove the dependency here?
 https://github.com/coq/coq/blob/master/intf/evar_kinds.ml#L23

> The Public/Internal thing loses the information of what functions are used by plugins in favour of giving them all Public stuff doesn't it?

Well, it only loses information if we want to. We can give plugins whatever subset of the API we want. How to organize that, I'm not sure.

> Would we make any promises of stability in the Public part or something? I still don't really understand what goes in Internal beyond renaming Unsafe to Internal.

I see more the distinction at the level of "this interface is supported" vs "this interface is not stable" etc... anyways API stability is a farce, there will be always be changed, but at least with [@@deprecated and modules we can manage them better.

**Comment by ejgallego** [comment]
> Umm, how does #6092 remove the dependency here?

Never mind, I see it now, you are adding `global_reference` to the kernel. That certainly needs review.

**Comment by maximedenes** [comment]
@ejgallego I have the feeling that the series of PRs is not easy to review. At least, @gares and I are struggling a bit because we don't fully see the global picture.

Concrete example: I was not expecting that plugins would be provided a `Constr` based interface. I thought instead they should use `EConstr`. But didn't you provide a `Public` interface in `Constr`? Maybe not, I didn't have time to look at the diffs today.

**Comment by ejgallego** [comment]
> @ejgallego I have the feeling that the series of PRs is not easy to review. At least, @gares and I are struggling a bit because we don't fully see the global picture.

The global picture is explained here not in the PRs, the PRs don't make still sense until we agree on a plan on the 3 different axis.

> Concrete example: I was not expecting that plugins would be provided a Constr based interface. I thought instead they should use EConstr. But didn't you provide a Public interface in Constr? Maybe not, I didn't have time to look at the diffs today.

Wow, I am very puzzled. Maybe we are looking at a different codebase? Are you aware of what you are proposing vs were the code, the plugins, and the _API_ are today?

**Comment by maximedenes** [comment]
> Wow, I am very puzzled. Maybe we are looking at a different codebase? Are you aware of what you are proposing vs were the code, the plugins, and the API are today?

Well, maybe I completely misunderstood @ppedrot 's plan, in which case he will correct me. But I thought using `Constr` from plugins was deprecated. Last time you told me you would not put deprecated stuff in `Public` interfaces, but rather have such plugins `open Internal`. Now again, maybe I misunderstood that...

Maybe this is showing an example where module boundaries are not the same as the API we want to expose to plugins and I find it interesting to see how we resolve that.

**Comment by ejgallego** [comment]
> But I thought using Constr from plugins was deprecated.

I'd dare to say that is not correct; or at least, I don't see how it can happen without rewriting the whole codebase. On the other hand, large parts of the system have not been yet ported to `EConstr` and here we are talking about _man years_ of work. It also seems strange that such demanding goals appear now, however code with way more basic problems got merged without a problem.

More generally, I feel that here we a hitting again the same issues that led to the current "API disaster", in particular, it seems we have a really bad definition of what the goals are and what we are trying to achieve.

My goal is very easy, and much more modest: I want to incorporate the "cleanups" [mainly deprecations] that happened in `API` into the main codebase, and remove all the hideous interface duplication while still maintaining some encapsulation.

**Comment by ejgallego** [comment]
More generally, we are worrying about "API", but there are very serious issues in the main codebase that nobody seems to care of. As we say in Spain, "parece que vamos a construir la casa por el tejado".

Many plugins just copy code from Coq, and of course they do the wrong thing.

**Comment by SkySkimmer** [comment]
I think for instance we don't want plugins to declare constants with econstr (and there's no interface for doing that). So for instance a program translation plugin would never see the shadow of an econstr or of an evar map. We can also see eg extraction doing strange things with econstr and Evd.empty that seem like they'd work better on constr.

>More generally, we are worrying about "API", but there are very serious issues in the main codebase that nobody seems to care of. As we say in Spain, "parece que vamos a construir la casa por el tejado".

Do we have github issues for these issues?

**Comment by ejgallego** [comment]
> Do we have github issues for these issues?

A few of them, not all. I have myself quite a few trees trying to solve the ones I care about, but that has been delayed, you guess by what :) by the API, the hugest time sink since Diablo II.

In particular, the `Constr/EConstr` port seems far from finished at the higher layers of the hierarchy, the proof engine / tactics port seems also with some TODOs and the STM is far from being well-integrated into the system. Not to talk about the `start/save_proof` path.

The more we talk the more I find this API project totally devoid of interesting, useful, and technical content [with pluginization coming as a second close for me] So far Coq has seen a net negative benefit from the API work.

**Comment by maximedenes** [comment]
> More generally, I feel that here we a hitting again the same issues that led to the current "API disaster", in particular, it seems we have a really bad definition of what the goals are and what we are trying to achieve.

Indeed, and that's what I was asking about.

> My goal is very easy, and much more modest: I want to incorporate the "cleanups" [mainly deprecations] that happened in API into the main codebase, and remove all the hideous interface duplication while still maintaining some encapsulation.

Sounds good. I don' t see why we call it API then, but those goals look very reasonable to me. In fact, I had explicitly told Matej not to deprecate stuff in the API, and that we should first clean the codebase instead. I don't know why that was ignored.

To move forward, I guess we need to make sure that developers agree with those two goals and the way these PRs implement them. Merging this week should be good, because the rest of the codebase is fairly stable these days.

**Comment by ppedrot** [comment]
> But I thought using Constr from plugins was deprecated.

No, `Constr` is meant to stay. It is just a different data structure than `EConstr`, which does not feature the same interface and is not designed for the same goals. Its main purpose is to describe kernel terms. For instance, the parametricity plugin is morally manipulating Constr rather than EConstr. (In practice it does not because it also allows to use tactics, but morally the pure translation part could be done only with kernel-side terms.)

**Comment by maximedenes** [comment]
> I think for instance we don't want plugins to declare constants with econstr (and there's no interface for doing that). So for instance a program translation plugin would never see the shadow of an econstr or of an evar map. We can also see eg extraction doing strange things with econstr and Evd.empty that seem like they'd work better on constr.

> No, Constr is meant to stay. It is just a different data structure than EConstr, which does not feature the same interface and is not designed for the same goals. Its main purpose is to describe kernel terms. For instance, the parametricity plugin is morally manipulating Constr rather than EConstr. (In practice it does not because it also allows to use tactics, but morally the pure translation part could be done only with kernel-side terms.)

Thanks guys for the clarification.

**Comment by ppedrot** [comment]
> More generally, I feel that here we a hitting again the same issues that led to the current "API disaster", in particular, it seems we have a really bad definition of what the goals are and what we are trying to achieve.

@ejgallego Unfortunately, I have to concur. And also, I think we're wasting precious time resources for an at best fuzzy result. I definitely believe there are way more important things to do, including ones that you might probably be the only one being able to work on, like *cough* the STM *cough*.

(That said, Diablo II is a pleasant time sink. Maybe you enjoy the API alike?)

**Comment by ejgallego** [comment]
> (That said, Diablo II is a pleasant time sink. Maybe you enjoy the API alike?)

Indeed it has a very similar feeling of repetitiveness. Ok, so let's get back on topic, I think we can start by merging #6064 and #6065. Right?

**Comment by maximedenes** [comment]
Indeed, I just "approved"  those two, they look ready to merge modulo small rebase.

**Comment by ejgallego** [comment]
Can you merge #6064 so I rebase #6065 on top of master? [It also needs a couple of fixes]

**Comment by maximedenes** [comment]
Done!

**Comment by ejgallego** [comment]
Ok so given that #6064 / #6065 are pretty uncontroversial, now we can go back to our original topic.

What should we do with `API.mli`. We clearly cannot ship that mess in a release. However, if we plan to control the interfaces, it has to be done somewhere. Recapitulating:

- Option A: remove API.mli . It works well for ostriches but puts us back at the beginning.
- Option B: keep API.mli , but export the full underlying interfaces. We at least keep one layer for future use, maintenance cost is low.
- Option C: keep API.mli (or kernel_API.mli) etc... and export only curated interfaces.

**Comment by ejgallego** [comment]
> Done!

Mersi !

**Comment by ejgallego** [comment]
Ok #6065 should be ready to go, Travis-allowing. So what I am going to do is to extract the deprecation-only part from #6084 and submit it separately. Ideally this should go in quick, as it will disrupt eveyone.

The good news is that once that new PR is in, we have extracted practically all the useful "duplication removal" efforts from the API, so concurrency of development could be high again.

**Comment by ppedrot** [comment]
@ejgallego So I stand my ground and will advocate for an attribute-based solution. I actually have a quite stupid proposal: we can simply piggy-back on the deprecated attribute and use it for internal functions, to the exception that we take advantage of the fact that it has a text payload. So essentially we set in the payload some characteristic message stating that this function is internal.

This allows for an easy cohabitation of the public and internal code, and with a bit of mind-bending, we can even think of our custom attribute as a proper one. I agree it is a hack and not very aesthetic, but maybe a few scripts would help.

**Comment by ejgallego** [comment]
Sure feel free to propose a patch doing that, however I think that using the deprecated attribute will be a super mess [we compile `-w error`].

You can use a custom attribute, but then you'll have nightmares to do something useful with it. It is hard to beat the tradeoff modules offer wrt namespacing.

**Comment by ejgallego** [comment]
Modules are also hard to beat wrt documentation and browsing.

Well, I've submitted #6098 , that should close the circle wrt moving deprecation in the API to the core. [I hope that PR works well with 4.06.0].

Once #6065 and #6098 are merged, to which I don't see a blocker, all the other prs are much smaller.

Likely we could close this issue by adopting option B above + some more refactoring, maybe leaving names with an Internal/Public part so it doesn't get lost.

Honestly we have advanced a lot and even if some hooks are still present in `library`, we are close to a  modular system. Both ocamlbuild and jbuilder are almost happy with Coq and indeed it looks like jbuilder may be faster for builds.

**Comment by ejgallego** [comment]
> given that plugins go through the API by default (which I protested against, fact that I will continue to remind), we really have to do something about it.

Packing things into modules also solve this problem nicely; if the plugin writer wants to opt-in, just do `open Kernel_API` and load the supported stuff. YMMV anyways.

**Comment by ejgallego** [comment]
Ok, so I've splitted the patches removing interface duplication from the core to #6065 #6098. These should be ready to merge. #6100 should be also merged.

After those are in, the value of `API.mli` is reduced to a simple subset of functions, that as @maximedenes said doesn't really have a lot of value, thou it has some.

Now the proposal to get rid of the redundancy while at the same time preserving some kind of interface is at thttps://github.com/coq/coq/pull/6066/commits/7e2d5f9691b522a517a6abc2d42394eeb532fc11 and is much easier to review; it is also fairly orthogonal to core development so we could take our time to think about it.

**Comment by Zimmi48** [comment]
I have a good general feeling about the approach you are proposing, although it will get easier to review once the first batch of consensual clean-up PRs get merged.

I see you are keeping the `API.mli` machinery while solving the code duplication issue. Is this approach also going to solve the problems @ppedrot and some others were experimenting with `-bypass-api` or not?

**Comment by ejgallego** [comment]
> I have a good general feeling about the approach you are proposing, although it will get easier to review once the first batch of consensual clean-up PRs get merged.

I am rebasing the PRs so now the "API" part should be in a small commit.

> I see you are keeping the API.mli machinery while solving the code duplication issue. Is this approach also going to solve the problems @ppedrot and some others were experimenting with -bypass-api or not?

I think it would. More importantly, keeping the `API.mli` file gives a lot of flexibility to try out approaches. For example, if I want to see how STM is used by the plugins, I just need to modify the module declaration in `API` and see what happens. That takes a few seconds.

**Comment by mattam82** [comment]
I wish there was a lightweight attribute-based solution as @ppedrot advocates... Option B "We at least keep one layer for future use, maintenance cost is low.", can you explain a bit more?

**Comment by ejgallego** [comment]
> I wish there was a lightweight attribute-based solution as @ppedrot advocates...

Umm, well, that would be hard to implement, and honestly I don't see the advantages of using attributes vs regular namespace. 

> "We at least keep one layer for future use, maintenance cost is low.", can you explain a bit more?

We just keep and `API.mli` file reexporting the base modules, but without carrying any special interface.

Basically that would be:
```
module Names = include module type of Names
module Univ = include module type of Univ
...
```
so we preserve the structure for future changes.

**Comment by ppedrot** [comment]
Note that I do believe that @ejgallego's proposal is orders of magnitude better than what we have currently. So, we can probably try this approach altogether. If that works, fine, otherwise, better than now.

**Comment by psteckler** [comment]
Is there a way to combine the deprecated-attribute and .mli file
approaches? I mean, putting the @@deprecated tags in an .mli file, so that
you easily see what's OK / not OK to use, rather than sprinkling those tags
around the whole code base?

**Comment by ejgallego** [comment]
Deprecated != not OK to use.

**Comment by psteckler** [comment]
> Deprecated != not OK to use / internal.

There's also `@@warning`, which might be better-suited.

**Comment by ejgallego** [comment]
> There's also @@warning, which might be better-suited.

What is `@@warning` and why is that better suited?

**Comment by psteckler** [comment]
`@@warning` is another OCaml annotation, and issues a warning to the user. It's better-suited than `@@deprecated`, because is it doesn't come with the semantic burden associated with that tag.

**Comment by ejgallego** [comment]
> @@warning is another OCaml annotation, and issues a warning to the user. It's better-suited than @@deprecated, because is it doesn't come with the semantic burden associated with that tag.

Care to show a working example?

**Comment by psteckler** [comment]
I meant `@@ppwarning`:

```let f x = x + 42 [@ppwarning "You're not a Coq dev"]```

**Comment by ejgallego** [comment]
> I meant `@@ppwarning:`
> 
>` let f x = x + 42 [@ppwarning "You're not a Coq dev"]`

Ah, OK. I don't see how that is very useful to us thou, it is almost the same as deprecated.

**Comment by herbelin** [comment]
Could someone be kind to summarize the status of this issue? I thought at some time that the maintenance of the API was going to be dispatched within the individual modules, so as to avoid duplication, but I may have misunderstood.

**Comment by ejgallego** [comment]
@herbelin basically the plan is to give a summary at the December's WG and then try agree on a plan.

**Comment by herbelin** [comment]
@ejgallego: ok.

**Comment by maximedenes** [comment]
I think we can close this, since there is no `API.mli` file anymore.
